<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [28]:
import pandas as pd
import psycopg2
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [29]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [30]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [147]:
# текст запроса
query_3_1 = f'''SELECT
                COUNT(DISTINCT id)
                FROM vacancies
            '''
df_3_1=pd.read_sql_query(query_3_1,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_21308\1231754512.py:6: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [148]:
# результат запроса
df_3_1

count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [149]:
# текст запроса
query_3_2 = f'''SELECT
                COUNT(DISTINCT id)
                FROM employers
            '''
df_3_2=pd.read_sql_query(query_3_2,connection)


C:\Users\админ\AppData\Local\Temp\ipykernel_21308\426972510.py:6: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [150]:
# результат запроса
df_3_2

count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [151]:
# текст запроса
query_3_3 = f'''SELECT
                COUNT(DISTINCT id)
                FROM areas
            '''
df_3_3=pd.read_sql_query(query_3_3,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_21308\717369303.py:6: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [152]:
# результат запроса
df_3_3

count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [153]:
# текст запроса
query_3_4 = f'''SELECT
                COUNT(DISTINCT id)
                FROM industries
            '''
df_3_4=pd.read_sql_query(query_3_4,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_21308\3021971014.py:6: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [154]:
# результат запроса
df_3_4

count
0    294

***

# Выводы по предварительному анализу данных:
1. Вакансий - 49197.
2. Работодателей - 23501.
3. Количество регионов - 1362.
4. Количество сфер деятельности 294.                                                                                                                         
Это достаточно общие данные.  Чтобы лучше оценить рынок труда нужен более детальный анализ. В целом видно что достаточно много вакансий. Скорее всего большая их чать сосредоточена в крупных городах, но это мы выясним далее. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [9]:
# текст запроса
query_4_1 = f'''SELECT
                    a.name as city,
                    COUNT(v.id) as vacancies
                FROM public.vacancies v
                JOIN public.areas a ON a.id=v.area_id
                GROUP BY a.name
                ORDER BY 2 DESC 
                LIMIT 5
            '''
df_4_1=pd.read_sql_query(query_4_1,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\2066056768.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1=pd.read_sql_query(query_4_1,connection)


In [10]:
# результат запроса
df_4_1

city  vacancies
0           Москва       5333
1  Санкт-Петербург       2851
2            Минск       2112
3      Новосибирск       2006
4           Алматы       1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [11]:
# текст запроса
query_4_2 = f'''SELECT
                    COUNT(id)
                FROM public.vacancies 
                WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
            '''
df_4_2=pd.read_sql_query(query_4_2,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\3568453820.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2=pd.read_sql_query(query_4_2,connection)


In [12]:
# результат запроса
df_4_2

count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [13]:
# текст запроса
query_4_3 = f'''SELECT
                    ROUND(AVG(salary_to)) avg_salary_to,
                    ROUND(AVG(salary_from)) avg_salary_from
                FROM public.vacancies 
            '''
df_4_3=pd.read_sql_query(query_4_3,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\862720138.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3=pd.read_sql_query(query_4_3,connection)


In [14]:
# результат запроса
df_4_3

avg_salary_to  avg_salary_from
0       110537.0          71065.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [15]:
# текст запроса
query_4_4 = f'''SELECT
                    schedule,
                    employment,
                    COUNT(id) as vacancies
                FROM public.vacancies
                GROUP BY schedule, employment
                ORDER BY COUNT(id) DESC
            '''
df_4_4=pd.read_sql_query(query_4_4,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\1675756839.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4=pd.read_sql_query(query_4_4,connection)


In [16]:
# результат запроса
df_4_4

schedule           employment  vacancies
0        Полный день     Полная занятость      35367
1   Удаленная работа     Полная занятость       7802
2      Гибкий график     Полная занятость       1593
3   Удаленная работа  Частичная занятость       1312
4     Сменный график     Полная занятость        940
5        Полный день           Стажировка        569
6     Вахтовый метод     Полная занятость        367
7        Полный день  Частичная занятость        347
8      Гибкий график  Частичная занятость        312
9        Полный день     Проектная работа        141
10  Удаленная работа     Проектная работа        133
11     Гибкий график           Стажировка        116
12    Сменный график  Частичная занятость        101
13  Удаленная работа           Стажировка         64
14     Гибкий график     Проектная работа         18
15    Сменный график           Стажировка         12
16    Вахтовый метод     Проектная работа          2
17    Сменный график     Проектная работа          1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [17]:
# текст запроса
query_4_5 = f'''SELECT
                    experience,
                    COUNT(id) as vacancies
                FROM public.vacancies
                GROUP BY experience
                ORDER BY COUNT(id) 
            '''
df_4_5=pd.read_sql_query(query_4_5,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\2370698125.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5=pd.read_sql_query(query_4_5,connection)


In [18]:
# результат запроса
df_4_5

experience  vacancies
0         Более 6 лет       1337
1           Нет опыта       7197
2       От 3 до 6 лет      14511
3  От 1 года до 3 лет      26152

***

# Выводы по детальному анализу вакансий
1. Наибольшее количество вакансий конечно же в Москве (5333 вакансии), на втором месте Санкт-Петербург (2851 вакансия), на третьем, четвертом и пятом местах Минск, Новосибирск и Алматы соответсвенно (на эти города приходится около 2000 вакансий на каждый). Проиллюстрируем эти данные для наглядности столбчатой диаграммой "Зависимость вакансий от региона".
2. У нас есть неполные данные по заработной плате. Почти в половине вакансий не указана либо верхняя либо нижняя граница зарплатной "вилки". В среднем нижняя граница заработной платы сотавляет 71065 руб, верхняя - 110537 рублей. 
3. Как зависит количество вакансий от рабочего графика и типа трудоустройства мы можем увидеть на многоуровневой столбчатой диаграмме "Зависимость количества вакансий от графика работы и типа занятости". Здесь мы видим что работодателей интересуют соискатели с полным рабочим днем и с полной занятостью, таких вакансий около 70%. На втором месте вакансии с удаленной работой и полной занятостью - 7802 вакансии. Гибкий график-Полная занятость на третьем месте (1593 вакансии), на четвертом месте: Удаленная работа-Частичная занятость (1312 вакансии). Самые редкие вакансии - это вакинсии со сменным графиком или вахтовым метом и проектной работой. 
4. Так же можно увидеть как зависит количество вакансий от опыта соискателя. Диаграмма 'Зависимость количества вакансий от опыта работы' очень хорошо это иллюстрирует. Наибольшее количество вакансий с опытом от 1 года до 3 лет (26152 вакансий). Далее идет опыт работы от 3 до 6 лет (14511 вакансий). Достаточное количество вакансий, где не требуется опыт работы (7197). И 1337 вакансий для соискателей с опытом работы более 6 лет. Как правило опыт работы от 1 до 3 лет - это начинающие специалисты и как правило эти вакансии предполагают большую текучку кадров, отсюда и такое огромное количество предлагаемых вакансий. Что- то похожее и с опытом от 3 до 6 лет, но уде в меньшей степени. Опыт работы более 6 лет - это уже специалисты, топ-менеджеры, которые как правило определились со сферой деятельности и текучка кадров в данном сегменте небольшая, отсюда и самая маленькая цифра по вакансиям. 


In [19]:
# строим столбчатую диаграмму
fig = px.bar(
    data_frame = df_4_1,
    x = 'city',
    y = 'vacancies',
    title = 'Зависимость вакансий от региона',
)
fig.show()
fig.write_html("plotly/bar_4_1.html")

In [20]:
# строим многоуровневую столбчатую диаграмму
fig = px.bar(
    data_frame = df_4_4,
    x = 'vacancies',
    y = 'schedule',
    barmode = "group",
    color = 'employment',
    orientation = 'h',
    title = 'Зависимость количества вакансий от графика работы и типа занятости'
)
fig.show()
fig.write_html("plotly/bar_4_4.html")

In [21]:
# строим многоуровневую столбчатую диаграмму
fig = px.bar(
    data_frame = df_4_5,
    x = 'vacancies',
    #y = 'experience',
    barmode = "group",
    color = 'experience',
    orientation = 'h',
    title = 'Зависимость количества вакансий от опыта работы'
)
fig.show()
fig.write_html("plotly/bar_4_5.html")

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [46]:
# текст запроса
query_5_1 = f'''SELECT
                    e.name as company,
                    COUNT(v.id) as vacancies
                FROM public.vacancies v
                JOIN public.employers e ON e.id=v.employer_id
                GROUP BY e.name
                ORDER BY COUNT(v.id) DESC
                LIMIT 5
 
            '''
df_5_1=pd.read_sql_query(query_5_1,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\2889411308.py:12: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [47]:
#результат запроса
df_5_1

company  vacancies
0         Яндекс       1933
1     Ростелеком        491
2       Тинькофф        444
3           СБЕР        428
4  Газпром нефть        331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [48]:
# текст запроса
query_5_2 = f'''SELECT 
                    a.name as area, 
                    COUNT(e.id) cnt_employer, 
                    COUNT(v.id) cnt_vacancies
                FROM public.areas a
                LEFT JOIN public.vacancies v ON a.id=v.area_id 
                LEFT JOIN public.employers e ON a.id=e.area
                GROUP BY a.name 
                HAVING COUNT(v.id) < 1 
                ORDER BY 2 DESC
                
 
            '''
df_5_2=pd.read_sql_query(query_5_2,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\2637552257.py:15: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [49]:
# результат запроса
df_5_2

area  cnt_employer  cnt_vacancies
0                          Россия           410              0
1                       Казахстан           207              0
2              Московская область            75              0
3              Краснодарский край            19              0
4              Ростовская область            18              0
..                            ...           ...            ...
588                      Струнино             0              0
589    Славгород (Алтайский край)             0              0
590                     Сорочинск             0              0
591  Ленинский (Тульская область)             0              0
592                         Тавда             0              0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [50]:
# текст запроса
query_5_3 = f'''SELECT
                    e.name as company,
                    COUNT(DISTINCT v.area_id) as area
                FROM public.employers e
                JOIN public.vacancies v ON e.id=v.employer_id
                GROUP BY e.name
                ORDER by 2 desc
 
            '''
df_5_3=pd.read_sql_query(query_5_3,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\1273211672.py:11: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [51]:
# результат запроса
df_5_3

company  area
0                      Яндекс   181
1                  Ростелеком   152
2                  Спецремонт   116
3      Поляков Денис Иванович    88
4                   ООО ЕФИН     71
...                       ...   ...
14761                  UniSol     1
14762            UNISTORY LLC     1
14763                   UNIT6     1
14764     United Distribution     1
14765              UNITEL LLC     1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [52]:
# текст запроса
query_5_4 = f'''SELECT 
                    COUNT(e.id) as employers
                FROM public.employers e
                LEFT JOIN public.employers_industries ei ON e.id=ei.employer_id
                LEFT JOIN public.industries i ON ei.industry_id=i.id
                WHERE i.name IS NULL
 
            '''
df_5_4=pd.read_sql_query(query_5_4,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\2790068464.py:10: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [53]:
# результат запроса
df_5_4

employers
0       8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [54]:
# текст запроса
query_5_5 = f'''SELECT 
                    e.name  as company
                FROM public.employers e
                LEFT JOIN public.employers_industries ei ON e.id=ei.employer_id
                GROUP BY e.name
                HAVING COUNT(ei.industry_id)=4
                ORDER BY e.name
                OFFSET 2
                LIMIT 1
            '''
df_5_5=pd.read_sql_query(query_5_5,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\1213728283.py:12: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [55]:
# результат запроса
df_5_5

company
0    2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [57]:
# текст запроса
query_5_6 = f'''SELECT
                    COUNT(ei.employer_id)
                FROM public.employers_industries ei
                JOIN public.industries i ON ei.industry_id=i.id
                WHERE i.name LIKE '%Разработка программного обеспечения%'
 
            '''
df_5_6=pd.read_sql_query(query_5_6,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\2884876586.py:9: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [58]:
# результат запроса
df_5_6

count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [44]:
# код для получения списка городов-милионников
query_5_7 = f'''(SELECT
                    a.name as city,
                    COUNT(v.id) as vacancies
                FROM public.areas a                                                                                                        
                JOIN public.vacancies v ON a.id=v.area_id
                JOIN public.employers e ON e.id=v.employer_id
                WHERE e.name = 'Яндекс' 
                AND a.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 
                'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск',
                'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 
                'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')                                                                                          
                GROUP BY a.name 
                ORDER BY 2)
                                                                                            
                UNION ALL
                                                                                            
                SELECT 
                    'Total',
                    COUNT(*)
                FROM public.areas a
                JOIN public.vacancies v ON a.id=v.area_id
                JOIN public.employers e ON e.id=v.employer_id
                WHERE e.name = 'Яндекс' 
                AND a.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 
                'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск',
                'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 
                'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
                

 
            '''
df_5_7=pd.read_sql_query(query_5_7,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\3991518015.py:33: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [45]:
# результат запроса
df_5_7

city  vacancies
0             Total        485
1              Омск         21
2         Челябинск         22
3        Красноярск         23
4         Волгоград         24
5             Пермь         25
6            Казань         25
7    Ростов-на-Дону         25
8               Уфа         26
9            Самара         26
10        Краснодар         30
11          Воронеж         32
12      Новосибирск         35
13  Нижний Новгород         36
14     Екатеринбург         39
15  Санкт-Петербург         42
16           Москва         54

***

# Выводы по анализу работодателей:
1. По количеству вакансий лидирует компания "Яндекс" (1933 вакансии). "Ростелеком" (491 вакансия), "Тинькофф" (444 вакансии), "СБЕР" (428 вакансии) на  втором, третьем и четвертом местах соответственно. Наглядно мы может это увидеть на диаграмме "Зависимость вакансий от компании работодателя"
2. Наибольшее количество работодателей находится в России.
3. Компания "Яндекс" публикуется в наибольшем количестве регионов (181), на втором месте компания "Ростелеком".
4. Наибольшее количество вакансий компании "Яндекс" приходится на Москву (54), на втором месте Санкт-Петербург (42), третье место - Екатеринбург (39), в остальных городах-миллионниках диапазон вакансий от 20 до 35.
Как мы видим особо выделяется компания "Яндекс". Создается впечатление, что либо в компании "Яндекс" постоянная текучка кадров, либо был открыт какой то крупный проект и во всех филиалах идет донабор сотрудников. Но я склоняюсь скорее к первому варианту.

In [59]:
# строим столбчатую диаграмму
fig = px.bar(
    data_frame = df_5_1,
    x = 'company',
    y = 'vacancies',
    title = 'Зависимость вакансий от компании работодателя',
)
fig.show()
# сохранение
fig.write_html("plotly/df_5_1.html")

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [61]:
# текст запроса
query_6_1 = f'''SELECT
                    COUNT(v.id)
                FROM public.vacancies v
                WHERE v.name ILIKE '%data%' OR v.name ILIKE '%данн%'
            '''
df_6_1=pd.read_sql_query(query_6_1,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\1456126991.py:7: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [80]:
# результат запроса
df_6_1

count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [62]:
# текст запроса
query_6_2 = f'''SELECT
                    COUNT(id)
                FROM public.vacancies 
                WHERE ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%')  
                    OR (name ILIKE '%исследователь данных%') OR ((name LIKE '%ML%') 
                    AND (name NOT ILIKE '%HTML%')) OR (name ILIKE '%machine learning%') 
                    OR (name ILIKE '%машинн%обучен%') AND (key_skills IS NOT NULL))
                    AND (name ILIKE '%junior%' OR experience ILIKE '%Нет опыта%' OR employment ILIKE '%Стажировка%')
            '''
df_6_2=pd.read_sql_query(query_6_2,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\3893771319.py:11: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [63]:
# результат запроса
df_6_2

count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [64]:
# текст запроса
query_6_3 = f'''SELECT
                    COUNT(id)
                FROM public.vacancies 
                WHERE ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%')  
                OR (name ILIKE '%исследователь данных%') OR ((name LIKE '%ML%') 
                AND (name NOT ILIKE '%HTML%')) OR (name ILIKE '%machine learning%') 
                OR (name ILIKE '%машинн%обучен%') AND (key_skills IS NOT NULL))
                AND (key_skills ILIKE '%SQL%' OR key_skills ILIKE '%postgres%')
            '''
df_6_3=pd.read_sql_query(query_6_3,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\4031810968.py:11: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [65]:
# результат запроса
df_6_3

count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [66]:
# текст запроса
query_6_4 = f'''SELECT
                    COUNT(id)
                FROM public.vacancies 
                WHERE ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%')  
                    OR (name ILIKE '%исследователь данных%') OR ((name LIKE '%ML%') 
                    AND (name NOT ILIKE '%HTML%')) OR (name ILIKE '%machine learning%') 
                    OR (name ILIKE '%машинн%обучен%') AND (key_skills IS NOT NULL))
                    AND (key_skills ILIKE '%Python%')

            '''
df_6_4=pd.read_sql_query(query_6_4,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\663468085.py:12: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [67]:
# результат запроса
df_6_4

count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [68]:
# текст запроса
query_6_5 = f'''SELECT
                    ROUND(AVG(length(key_skills) - length(REPLACE(key_skills, CHR(9), '' ))), 2) + 1
                FROM public.vacancies 
                WHERE ((name ILIKE '%data scientist%') OR (name ILIKE '%data science%')  
        OR (name ILIKE '%исследователь данных%') OR ((name LIKE '%ML%') 
        AND (name NOT ILIKE '%HTML%')) OR (name ILIKE '%machine learning%') 
        OR (name ILIKE '%машинн%обучен%') AND (key_skills IS NOT NULL))


            '''
df_6_5=pd.read_sql_query(query_6_5,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\673571023.py:12: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [69]:
# результат запроса
df_6_5

?column?
0      6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [70]:
# текст запроса
query_6_6 = f'''SELECT
                    experience,
                    ROUND(AVG(COALESCE((salary_from + salary_to)/2, salary_from, salary_to))) salary_avg
                FROM public.vacancies
                WHERE ((name ILIKE '%data scientist%') 
                    OR (name ILIKE '%data science%')  
                    OR (name ILIKE '%исследователь данных%')    
                    OR ((name LIKE '%ML%') 
                    AND (name NOT ILIKE '%HTML%')) 
                    OR (name ILIKE '%machine learning%') 
                    OR (name ILIKE '%машинн%обучен%'))
                    AND ((salary_from IS NOT NULL) OR (salary_to IS NOT NULL))
                GROUP BY experience


            '''
df_6_6=pd.read_sql_query(query_6_6,connection)

C:\Users\админ\AppData\Local\Temp\ipykernel_4952\1883489396.py:18: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [71]:
# результат запроса
df_6_6

experience  salary_avg
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

# Выводы по предметному анализу:
1. 1771 вакансия имеет отношение к данным.
2. 51 вакансия есть для начинающих дата-сайентистам.
3. В 201 вакансии неибходимы SQL и Postgres как ключевые навыки.
4. В 351 вакансии необходим Python как ключевой навык.
5. В среднем указывают 6 ключевых навыков в вакансиях DS.
6. Самая высокая средняя заработная плата составляет 243115 рублей (опыт работы от 3 до 6 лет). У начинающих дата-саентистов средняя заработная плата может составлять 74643 рублей.

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования